ARROW-4552: [JS] Add high-level Table and Column convenience methods

trxcllnt opened the pull request:
This PR closes the following JIRAs:

* [ARROW-4552](https://issues.apache.org/jira/browse/ARROW-4552) - Add Table and Schema `assign(other)` implementations
* [ARROW-2764](https://issues.apache.org/jira/browse/ARROW-2764) - Easy way to create a new Table with an additional column
* [ARROW-4553](https://issues.apache.org/jira/browse/ARROW-4553) - Implement Schema/Field/DataType comparators
* [ARROW-4554](https://issues.apache.org/jira/browse/ARROW-4554) - Implement logic for combining Vectors with different lengths/chunksizes
* [ARROW-4555](https://issues.apache.org/jira/browse/ARROW-4555) - Add high-level Table and Column creation methods
* [ARROW-4557](https://issues.apache.org/jira/browse/ARROW-4557) - Add Table/Schema/RecordBatch `selectAt(...indices)` method

I extracted a few more  high-level helper methods I've had laying around for creating, selecting, or manipulating Tables/Columns/Schemas/RecordBatches.

1. We currently have a `table.select(...colNames)` implementation, so I also added a `table.selectAt(...colIndices)` method to complement. Super handy when you have duplicates.
2. I added a basic `table.assign(otherTable)` impl. I added logic to compare Schemas/Fields/DataTypes in order to de-dupe reliably, which lives in the [`TypeComparator` Visitor](https://github.com/trxcllnt/arrow/blob/a67bd562cf6c4860bdce027981df859398e41b6d/js/src/visitor/typecomparator.ts#L83). I expose this via `compareTo()` methods on the Schema, Field, and DataType for ease of use. Bonus: the Writer [can now discern](https://github.com/trxcllnt/arrow/blob/a67bd562cf6c4860bdce027981df859398e41b6d/js/src/ipc/writer.ts#L129) between RecordBatches of the same stream whose Schemas aren't reference-equal.
3. I've also added logic to distribute Vectors of different lengths (or different internal chunk sizes) evenly across RecordBatches, to support a nearly zero-copy `Table#assign()` impl. I say nearly zero-copy, because there's a bit of allocation/copying to backfill null bitmaps if chunks don't exactly line up. But this also means [it's a bit easier](https://github.com/trxcllnt/arrow/blob/a67bd562cf6c4860bdce027981df859398e41b6d/js/test/unit/table-tests.ts#L178) now to create Tables or RecordBatches from values in-memory whose lengths may not exactly line up:
```ts
const table = Table.new(
  Column.new('foo', IntVector.from(arange(new Int32Array(10))),
  Column.new('bar', FloatVector.from(arange(new Float32Array(100))))
);
```
4. And lastly, I added [some more more tests](https://github.com/trxcllnt/arrow/blob/js/high-level-table-column-fns/js/test/unit/table/serialize-tests.ts#L38) to ensure various combinations of select/slice/concat/assign can round-trip through IPC and back again.

```ts

const table1 = Table.new(
    Column.new('a', Int32Vector.from(i32s)),
    Column.new('b', Float32Vector.from(f32s)),
    Column.new('c', Float64Vector.from(f64s))
);

const table2 = Table.new(
    Column.new('d', Utf8Vector.from(strs)),
    Column.new('d', BoolVector.from(bools)),
    Column.new('d', Int32Vector.from(i32s)),
);

const table3 = table1.select('b', 'c').assign(table2.selectAt(0, 1));

console.log(table3.schema.fields)
// > [
// >     ('b', Float32), 
// >     ('c', Float64), 
// >     ('d', Utf8), 
// >     ('d', Bool)
// > ]
```
(cc: @domoritz)

TheNeuralBit commented on js/src/table.ts:
```diff
@@ -93,15 +96,34 @@ export class Table<T extends { [key: string]: DataType; } = any>
-    public static fromVectors<T extends { [key: string]: DataType; } = any>(vectors: VType<T[keyof T]>[], names?: (keyof T)[]) {
-        return new Table(RecordBatch.from(vectors, names));
+    public static fromVectors<T extends { [key: string]: DataType; } = any>(vectors: Vector<T[keyof T]>[], fields?: (keyof T | Field<T[keyof T]>)[]) {
+        return Table.new<T>(vectors, fields);
```
I like the pattern where `from` methods are intended for conversions from other types and `new` methods are effectively constructors where we can use generics.

Do you think we should just go ahead and get rid of this to follow that pattern? or mark it deprecated or something? It's only been part of 0.4.0 for a couple of days I don't think anyone's using it.

trxcllnt replied:
I'm down for either. I couldn't remember if this method predated 0.4 or not, so I left it unchanged.

domoritz replied:
I think anyone would expect constructors instead of `.new` methods. Since we are pre 1.0 everything is fair game and you can be backwards incompatible in minor version bumps.

TheNeuralBit replied:
@domoritz agreed, I say let's just rename this (and `RecordBatch.from`) as an overload of `new`, we haven't even put out a blog post or anything touting these functions in 0.4.0. We can just tout the `new` versions instead.

TheNeuralBit commented on js/src/recordbatch.ts:
```diff
@@ -38,28 +38,27 @@ export class RecordBatch<T extends { [key: string]: DataType } = any>
-            vectors.reduce((len, vec) => Math.max(len, vec.length), 0),
-            vectors
-        );
+    public static from<T extends { [key: string]: DataType } = any>(chunks: (Data<T[keyof T]> | Vector<T[keyof T]>)[], names: (keyof T)[] = []) {
```
Maybe this should also be deprecated in favor of a `new` version?

trxcllnt commented:
> I'm a little concerned about allowing users to create Tables from different length vectors though. Is that something that's allowed in the other implementations? It feels like a case where the user is probably making a mistake and we should throw an error.

I don't agree that error-throwing would be appropriate here for two main reasons:
1. Practically speaking, aligning Column lengths before creating Tables is an unnecessary burden on library consumers. It involves first knowing or finding the max column length, then re-scanning all the columns to concatenate null chunks to the shorter ones. Presently this a non-trivial task (considering the internal Vector buffer layout) and/or would be slower to do externally (copies, suboptimal null bitmap creation, etc). I think it's quite a bit more neighborly of us to do this automatically.

1. On principle, node idioms suggest Errors should be used to communicate catastrophic unrecoverable failure modes, indicating program[mer] errors instead of logic errors. Tools like TS enforce most of these (e.g. it won't compile if you pass a string to a function that accepts numbers), but there are others (null reference, garbage data, etc.) that can happen at runtime, and error-throwing is appropriate then.

I'll clarify the behavior in this PR a bit further, since it's is really two features rolled into one.

First, there's the re-chunking of the Table based on the chunk layout of the Columns we're joining. This features says if we're combining two columns of equal length into a table, the table will contain as many RecordBatches as the child with the most chunks. This is 100% zero-copy, since it's only slicing the Vectors and distributing them evenly by lowest common chunk size:

```ts
a = Column.new('a', [1, 2, 3, 4, 5, 6])
b = Column.new('b', [1, 2, 3], [4, 5, 6])
assert(Table.new(a, b).chunks.length === 2)
```

I'm not certain whether it uses the same algorithm, but I think I recall @wesm mentioning the C++ implementation does support re-chunking.

The second part dealings with Columns of different lengths. For `assign()`, this means re-chunking up to the length of each column, then filling any remaining slots with nulls:
```ts
a = Column.new('a', [1, 2, 3, 4])
b = Column.new('b', [1, 2, 3], [4, 5, 6], [7, 8, 9])
assert(Table.new(a, b).chunks.length === 4)
// row_id |    a | b   ___
//      0 |    1 | 1      |
//      1 |    2 | 2      | <-- RecordBatch 1
//      2 |    3 | 3   ___|
//      3 |    4 | 4   ___| <-- RecordBatch 2
//      4 | null | 5      |
//      5 | null | 6   ___| <-- RecordBatch 3
//      3 | null | 7      |
//      4 | null | 8      | <-- RecordBatch 4
//      5 | null | 9   ___|
```

This example illustrates both features working in concert:
1. Column A is split into two chunks (sizes [3, 1]), because B's first chunk size is 3
2. Column A's second chunk has 1 data value, so Column B's _second_ chunk is split into two, since the LCD of A and B's second chunk is 1
3. Column B's third chunk is the leftover from the previous split, and its length is 2. Column A has no more chunks, so a null-filled chunk of length 2 is created as a placeholder
4. Column B's last chunk is left un-split, and another all-null chunk is created for A

TheNeuralBit commented:
Yeah I agree that the ability to pad short columns with nulls up to the length of the longest column is a nice bit of code to give our users when they need it, I'm just not sure it's always the right thing to do.  We could still give this functionality to users somehow without making it the default.

I really think most of the time someone is trying to create a Table from columns with an unequal length it's because they messed up somewhere, not because they want the shorter ones padded with nulls.

All that being said, I did some tests with pandas and pyarrow to see how they handle this. pandas will let you create a DataFrame from unequal length Series:
```
  In [14]: pd.DataFrame({'a': pd.Series([1,2,3]), 'b': pd.Series([1,2,3,4,5,6])})
Out[14]: 
     a  b
0  1.0  1
1  2.0  2
2  3.0  3
3  NaN  4
4  NaN  5
5  NaN  6
```

pyarrow does not, but pyarrow is lower-level than I think we are trying to target with arrow JS:
```
pa.RecordBatch.from_arrays([pa.Array.from_pandas(pd.Series([1,2,3])), pa.Array.from_pandas(pd.Series([1,2,3,4]))], ['a ', 'b'])
// ValueError
```

Given the precedent in pandas I could relent on this point.

trxcllnt commented:
@TheNeuralBit Yeah this behavior was inspired by pandas df.assign(). I do feel like padding with nulls is the friendliest way to go considering the alternatives -- if you accidentally combine some columns with uneven lengths and didn't mean to, you could always `table.slice()` after the fact to remove any unwanted rows from the end. Perhaps we should add `table.setColumnAt(idx, col)`, to make it easier to correct on a per-column basis:
```ts
let vals = [1, 2, 3, 4, 5, 6, 7, 8, 9]
let table = Table.new(Column.new('a', vals.slice(0, 4)),  Column.new('b', vals))
// correct the mistake
let a = table.getColumn('a').slice(0, 4)
let a2 = IntVector.new(vals.slice(4))
table.setColumn('a', a.concat(a2))
```

domoritz commented:
This still looks a bit odd to me

```
const table = Table.new(
  Column.new('foo', IntVector.from(arange(new Int32Array(10))),
  Column.new('bar', FloatVector.from(arange(new Float32Array(100))))
);
```

Why not

```
const table = new Table(
  new Column('foo', IntVector.from(arange(new Int32Array(10))),
  new Column('bar', FloatVector.from(arange(new Float32Array(100))))
);
```

domoritz commented:
About padding with `null`s, I think it could be good to show a warning when this padding is happening. I agree that it is very convenient but it can be surprising.

trxcllnt commented:
@domoritz I borrowed the `Ctor.new()` initialization pattern from `tensorflow-js`. It's a bit more flexible, since in TS constructors can't have generics of their own and can't be typed to return anything other than the exact type for which they're a constructor. I first used it in `Vector.new()` as an easy way to create the concrete Vector subclass for any Arrow DataType:

```
const data = Data.Int(new Int64(), ...rest));
const vec = Vector.new(data); // vec type is inferred to Int64Vector
assert(vec instanceof Int64Vector) // true
assert(typeof vec.toBigInt64Array == 'function') // has any methods specific to the Int64Vector
assert(vec.toBigInt64Array() instanceof BigInt64Array) // true
```

domoritz commented:
I see, that makes sense. I also think it's better not to have two different ways of initializing objects (constructors and `.new`) depending on whether you need to use generics in the constructor or not.

trxcllnt commented:
@domoritz I forgot to mention the distinction between `Ctor.from()` vs. `Ctor.new()`. `Ctor.new` is meant to be a more flexible constructor, since it's easier to define polymorphic argument lists and narrow the return signatures. `Ctor.from` is meant to initialize an instance from non-library types or data sources, which may require parsing, compute, or copies. An example in the core library is `Array.from()`, which takes Array-likes, Iterables, and an optional map function.

TheNeuralBit commented:
On the null-padding point: Maybe the right thing to do is to add a docstring and put a note about padding unequal length vectors.

Then at least people browsing the API docs would know it's happening.

domoritz commented:
What do you think about using `console.warn`?

TheNeuralBit commented:
I'd be fine with it if @trxcllnt is

trxcllnt commented:
:+1: for adding a note in the doc description

:-1: for `console.warn()` because it doesn't play nice with production log strategies. It's difficult in general for libraries to log output test-ably/cross-platform/efficiently without taking on external dependencies, exposing hooks, creating/maintaining custom debugging tools, or dictating something else about the user's setup.

Relying on globals like `console` forces people into awful workarounds like patching the `console` Object. [`debug`](https://www.npmjs.com/package/debug) also requires patching and doesn't produce logs that are easy to ingest. [`warning`](https://www.npmjs.com/package/warning) used by React expects the warnings to be compiled out via Webpack and would require us to start building/shipping dev versions of our UMD bundles.

domoritz commented:
Vega uses a tiny logger that users can override (https://github.com/vega/vega/blob/master/packages/vega-util/src/logger.js). Would it make sense for arrow to add a logger as well?

trxcllnt commented:
@domoritz it looks like that's using the global `console`, which causes pains for production log ingestion setups. In general I'd prefer not to reimplement logging in a dataframe library.

Furthermore, I'm not convinced this is even warn-level behavior since it's only allocating a few bitmaps. Pandas doesn't warn here either.

domoritz commented:
First, this is just the default logger and you can override it to print to something else. Second, I somewhat disagree that filling nulls when sizes mismatch is something that should go unnoticed. I think it is valuable to fill nulls but not showing any warnings will lead to errors that are hard to debug. I am okay with deferring this discussion but I feel that at some point arrow will need warnings.

trxcllnt commented:
@domoritz I agree with you on principle, we will probably want to log things. I do think doing it in a way that's easy to configure externally is important, and outside the scope of this PR. Could you add a JIRA epic so someone can pick it up after this is merged?

domoritz commented:
Here it is https://issues.apache.org/jira/browse/ARROW-4588

trxcllnt commented:
@TheNeuralBit ok, I removed `Table.fromVectors` added `RecordBatch.new`, and added the following description the `Table.new` tsdoc:
![screenshot from 2019-02-20 11-46-42](https://user-images.githubusercontent.com/178183/53119926-783ec900-3505-11e9-8856-2b0c5078f70f.png)

TheNeuralBit commented:
Perfect! thank you
